### PR TITLE
adding aquifer summary keywords

### PR DESCRIPTION
**Pull request by GitPaean**
AAQR, AAQT, AAQP

### CODE REVIEW

**Comment by GitPaean** [comment]
Hi, @bska , I looked into a little about these summary keywords, while again, I feel like the most efficient way to do this is to get some instructions from you. 

Especially, it is sort of new types of summary keywords. Maybe you will has some opinions about the correct way to do this.

**Comment by GitPaean** [comment]
The existing modification is just to create the PR, not meaning it is the suggested way yet.

**Comment by bska** [comment]
> I feel like the most efficient way to do this is to get some instructions from you.

Thanks for reaching out.  Here's my initial assessment.

1. The [`SummaryNode`](https://github.com/OPM/opm-common/blob/082ab35a0e65f406ce447c3bab791b4d6d7735ab/opm/io/eclipse/SummaryNode.hpp#L30) class already recognizes the [`Aquifer`](https://github.com/OPM/opm-common/blob/082ab35a0e65f406ce447c3bab791b4d6d7735ab/opm/io/eclipse/SummaryNode.hpp#L32) keyword category and the new keywords added here don't introduce new quantity types.  We already support [`Rate`, `Total`, and `Pressure`](https://github.com/OPM/opm-common/blob/082ab35a0e65f406ce447c3bab791b4d6d7735ab/opm/io/eclipse/SummaryNode.hpp#L43-L51) types.

2. That means that your primary task is to make the `SummaryConfig` class aware of aquifer keywords by creating a keyword parser for the `AAQ*` keywords.  That comes down to creating a function `keywordA()`&mdash;if you wish to keep with the existing naming conventions&mdash;and to hook that up in the category switch for both the [general keywords](https://github.com/OPM/opm-common/blob/082ab35a0e65f406ce447c3bab791b4d6d7735ab/src/opm/parser/eclipse/EclipseState/SummaryConfig/SummaryConfig.cpp#L773) and the ["meta" keywords](https://github.com/OPM/opm-common/blob/082ab35a0e65f406ce447c3bab791b4d6d7735ab/src/opm/parser/eclipse/EclipseState/SummaryConfig/SummaryConfig.cpp#L811).  The latter handles collections of keywords such as `ALL` or `FWMSET`.  Note, by the way, that the [`aquifer classifier`](https://github.com/OPM/opm-common/blob/082ab35a0e65f406ce447c3bab791b4d6d7735ab/src/opm/parser/eclipse/EclipseState/SummaryConfig/SummaryConfig.cpp#L254-L256) appears to misinterpret the collection `ALL` as an aquifer keyword.  We should either fix that or ensure that this function is never called with an input string of `ALL`.

3. The parser must create appropriate `SummaryConfigNode` instances and add those to the `list` of nodes that is managed by the `SummaryConfig` class.  Creating one of those nodes amounts to code like
    ```C++
    auto param = SummaryConfigNode {
        keyword.name(), SummaryConfigNode::Category::Segment, keyword.location()
    }
    .isUserDefined( is_udq(keyword.name()) );
    ```
    followed by a loop over the items of keyword data (if any) to capture the aquifer ID in the [`number()`](https://github.com/OPM/opm-common/blob/082ab35a0e65f406ce447c3bab791b4d6d7735ab/src/opm/parser/eclipse/EclipseState/SummaryConfig/SummaryConfig.cpp#L611) field.  If there is no keyword data, then there should be one `SummaryConfigNode` for each analytic aquifer.

4. Then you need to create evaluation functions similar to those you created for the [segment pressure values](#1705) recently.  Those will, in turn, require some dynamic data&mdash;typically calculated by the simulator&mdash;and sent to the output layer in some fashion.  If you don't want to re-architect the complete communication protocol/interface you can probably leverage the [aquifer](https://github.com/OPM/opm-common/blob/082ab35a0e65f406ce447c3bab791b4d6d7735ab/opm/output/eclipse/RestartValue.hpp#L74) data member of class `RestartValue` to move the value across the module boundary.  I added that member as an early experiment in #1277 to enable reading aquifer data from restart files.  I'm sure we can expand/rework the [`AquiferData`](https://github.com/OPM/opm-common/blob/082ab35a0e65f406ce447c3bab791b4d6d7735ab/opm/output/data/Aquifer.hpp#L46) structure to meet your needs here.

5. Finally you need to hook those evaluation functions up to the big [table](https://github.com/OPM/opm-common/blob/082ab35a0e65f406ce447c3bab791b4d6d7735ab/src/opm/output/eclipse/Summary.cpp#L882) keyed by the summary keyword in order to have the summary output system recognize the new support.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=2829

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=2829

**Comment by akva2** [comment]
jenkins build this opm-simulators=2829 please

**Comment by GitPaean** [comment]
Thanks @akva2 , I  did not see the problem.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=2829 please

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=2829 please

**Comment by GitPaean** [comment]
from my side, it is ready for review.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=2829 please

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=2829 update_data please

File: src/opm/parser/eclipse/EclipseState/AquiferCT.cpp [hunk 1]
```diff
@@ -175,4 +175,11 @@ const std::vector<AquiferCT::AQUCT_data>& AquiferCT::data() const {
     return this->m_aquct;
 }
 
+bool AquiferCT::hasAquifer(const int aquID) const {
+    const auto it = std::find_if(this->m_aquct.begin(), this->m_aquct.end(),
+                                 [&aquID](const auto& aqu){ return aqu.aquiferID == aquID; });
+
+    return ( it != this->m_aquct.end() );
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/AquiferCT.cpp:182]: Any reason to not use `std::any_of()` here?
```C++
return std::any_of(this->m_aquct.begin(), this->m_acuct.end(),
                   [acuID](const auto& aqu) { return aqu.aquiferID == aquID; });
```

File: src/opm/parser/eclipse/EclipseState/Aquifetp.cpp [hunk 2]
```diff
@@ -113,4 +113,11 @@ std::vector<Aquifetp::AQUFETP_data>::const_iterator Aquifetp::end() const {
     return this->m_aqufetp.end();
 }
 
+bool Aquifetp::hasAquifer(const int aquID) const {
+    const auto it = std::find_if(this->m_aqufetp.begin(), this->m_aqufetp.end(),
+                                 [&aquID](const auto& aqu){ return aqu.aquiferID == aquID; });
+
+    return ( it != this->m_aqufetp.end() );
```
Comment (bska) [inline src/opm/parser/eclipse/EclipseState/Aquifetp.cpp:120]: Same as for the CT case.  I believe `std::any_of()` is more expressive here.

File: opm/output/data/Aquifer.hpp [hunk 3]
```diff
@@ -46,14 +46,31 @@ namespace Opm { namespace data {
...
+        {
+            if ( key == "AAQR" ) {
+                return this->fluxRate;
+            } else if ( key == "AAQT" ) {
+                return this->volume;
+            } else if ( key == "AAQP" ) {
+                return this->pressure;
+            }
+            return 0.;
+        }
```
Comment (bska) [inline opm/output/data/Aquifer.hpp:68]: Long term, I don't think this should be a member function.  All the data members are already public, so the "string to member" mapping could be handled elsewhere.

**Review by bska — COMMENTED** [review]
Looks good for the most part, although there are aspects that could be simplified by using functions in the `<algorithms>` library.

File: src/opm/parser/eclipse/EclipseState/AquiferCT.cpp (hunk 1 above)
Comment (GitPaean) [reply src/opm/parser/eclipse/EclipseState/AquiferCT.cpp]: No other reason, I just has not used `std::any_of` before.

File: opm/output/data/Aquifer.hpp (hunk 3 above)
Comment (GitPaean) [reply opm/output/data/Aquifer.hpp:68]: Yes. I have the same impression. When I was struggling to figure out the data flow, here, I used the easiest way to get the result for testing. If it is okay to you, let us address it when we add more aquifer summary keywords next time, which is in the near future.

**Comment by GitPaean** [comment]
jenkins build this opm-simulators=2829 opm-tests=398 please

**Review by bska — APPROVED** [review]
Thanks for the updates.  I like this version and I'll merge into master.

**Comment by bska** [comment]
Updated reference solution is installed in master.  I'll merge this now.  Thanks a lot for getting us to this stage.

**Pull request merged**